First version of the raw radar data reader

antoniopug opened the pull request:
I tested the example script with acquisition `/mnt/zen-ffa/nyquist/2020-12-04_0048/000` and calibration file `ZenCode/ZenModel/radar/antenna/intrinsic/ZRCA0003.pt`.

The things I would appreciate looking at are:
- The readme instructions (do they work for you)?
- The documentation
- The example.py script: Is it easy to read and understand?
- The rest of python/proto code: I don't think anything here is proprietary, but it would be good to have another set of eyes.

antoniopug committed:
Add first version of raw radar data reader

antoniopug committed:
Updat table

osamasal commented on raw_radar_reader/README.md:
````diff
@@ -0,0 +1,72 @@
+struct
+protobuf
+argparse
+```
````
Shouldn't we provide installation steps or a ```requirements.txt``` file to help users install these packages? also, are there certain version numbers that users should install?

On my laptop machine (which is not setup for building ```ZenCode```, I get the following errors when trying to install gzip, os, and struct:

```
/home/osama/Code/ZendarSDK/raw_radar_reader/protocol>pip3 install gzip 
Collecting gzip
Exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 353, in run
    wb.build(autobuilding=True)
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 749, in build
    self.requirement_set.prepare_files(self.finder)
  File "/usr/lib/python3/dist-packages/pip/req/req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/usr/lib/python3/dist-packages/pip/req/req_set.py", line 554, in _prepare_file
    require_hashes
  File "/usr/lib/python3/dist-packages/pip/req/req_install.py", line 278, in populate_link
    self.link = finder.find_requirement(self, upgrade)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 465, in find_requirement
    all_candidates = self.find_all_candidates(req.name)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 423, in find_all_candidates
    for page in self._get_pages(url_locations, project_name):
  File "/usr/lib/python3/dist-packages/pip/index.py", line 568, in _get_pages
    page = self._get_page(location)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 683, in _get_page
    return HTMLPage.get_page(link, session=self.session)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 795, in get_page
    resp.raise_for_status()
  File "/usr/share/python-wheels/requests-2.18.4-py2.py3-none-any.whl/requests/models.py", line 935, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://pypi.org/simple/gzip/
/home/osama/Code/ZendarSDK/raw_radar_reader/protocol>pip3 install os
Collecting os
Exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 353, in run
    wb.build(autobuilding=True)
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 749, in build
    self.requirement_set.prepare_files(self.finder)
  File "/usr/lib/python3/dist-packages/pip/req/req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/usr/lib/python3/dist-packages/pip/req/req_set.py", line 554, in _prepare_file
    require_hashes
  File "/usr/lib/python3/dist-packages/pip/req/req_install.py", line 278, in populate_link
    self.link = finder.find_requirement(self, upgrade)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 465, in find_requirement
    all_candidates = self.find_all_candidates(req.name)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 423, in find_all_candidates
    for page in self._get_pages(url_locations, project_name):
  File "/usr/lib/python3/dist-packages/pip/index.py", line 568, in _get_pages
    page = self._get_page(location)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 683, in _get_page
    return HTMLPage.get_page(link, session=self.session)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 795, in get_page
    resp.raise_for_status()
  File "/usr/share/python-wheels/requests-2.18.4-py2.py3-none-any.whl/requests/models.py", line 935, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://pypi.org/simple/os/
/home/osama/Code/ZendarSDK/raw_radar_reader/protocol>pip3 install struct
Collecting struct
Exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 353, in run
    wb.build(autobuilding=True)
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 749, in build
    self.requirement_set.prepare_files(self.finder)
  File "/usr/lib/python3/dist-packages/pip/req/req_set.py", line 380, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/usr/lib/python3/dist-packages/pip/req/req_set.py", line 554, in _prepare_file
    require_hashes
  File "/usr/lib/python3/dist-packages/pip/req/req_install.py", line 278, in populate_link
    self.link = finder.find_requirement(self, upgrade)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 465, in find_requirement
    all_candidates = self.find_all_candidates(req.name)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 423, in find_all_candidates
    for page in self._get_pages(url_locations, project_name):
  File "/usr/lib/python3/dist-packages/pip/index.py", line 568, in _get_pages
    page = self._get_page(location)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 683, in _get_page
    return HTMLPage.get_page(link, session=self.session)
  File "/usr/lib/python3/dist-packages/pip/index.py", line 795, in get_page
    resp.raise_for_status()
  File "/usr/share/python-wheels/requests-2.18.4-py2.py3-none-any.whl/requests/models.py", line 935, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://pypi.org/simple/struct/
```

antoniopug replied:
Thanks for catching that, I added a requirements.txt. 

The packages you had trouble with are part of standard python, I neglected to prune them from my dependency list.

antoniopug replied:
Can you try again?

osamasal replied:
OK, here's what I'm getting right now after ignoring the packages I mentioned above:

```
/home/osama/Code/ZendarSDK/raw_radar_reader>python3 ./python/example.py --root ~/zen-ffa/nyquist/2020-12-04_0048/000 --calibration ~/zen-ffa/nyquist/2020-12-04_0048/000
Traceback (most recent call last):
  File "./python/example.py", line 5, in <module>
    from acquisition import AcquisitionStreamModel
  File "/home/osama/Code/ZendarSDK/raw_radar_reader/python/acquisition.py", line 5, in <module>
    from acquisition_pb2 import (
ModuleNotFoundError: No module named 'acquisition_pb2'
```
Looks like I have built this file under the protocol directory:
```
ZendarSDK/raw_radar_reader/protocol/acquisition_pb2.py
```

and my PYTHONPATH looks OK too:
```
PYTHONPATH=:/home/osama/code/ZendarSDK/raw_radar_reader/protocol:/home/osama/code/ZendarSDK/raw_radar_reader/python
```

Did I miss something?

antoniopug replied:
I cannot reproduce your issue. I started from a new terminal, zeroed out my PYTHONPATH and copy-pasted the instructions in the README and the command runs through just fine. 

Are you sure that `/home/osama/code/ZendarSDK/raw_radar_reader/protocol` and `/home/osama/code/ZendarSDK/raw_radar_reader/python` exist?

osamasal replied:
I just checked now, and these two directories exist on my machine. I also just retried your new ```requirements.txt``` file and the installation when smoothly this time. This seems to be resolved now.

osamasal commented on raw_radar_reader/README.md:
```diff
@@ -0,0 +1,72 @@
+cd ./protocol
+protoc primitive.proto --python_out .
+protoc radar.proto --python_out .
+protoc acquisition.proto --python_out .
```
If we're building inside the repro (not preferred), then we should provide instructions on how to undo the build output.

osamasal commented on raw_radar_reader/README.md:
````diff
@@ -0,0 +1,72 @@
+## Example Script
+
+```
+python3 ./python/example.py --root <path-to-data> --calibration <path-to-cal-file>
````
Is ```<path-to-data>``` the root directory (i.e. ```zen-ffa/nyquist/2020-12-04_0048/000```)?

antoniopug replied:
Yes

osamasal reviewed (commented):
I only made it a few steps into the instructions. I had some problems and comments, which I added to this PR.

jimmy-zendar commented on raw_radar_reader/README.md:
````diff
@@ -0,0 +1,72 @@
+
+To use these tools, perform the following steps:
+```
+cd ./protocol
````
can we create these protos the same way as we create the protos for the viewer?

antoniopug replied:
Isn't the way we create protos for the viewer to embed the deb file? I'd love to do that once it's ready and get rid of the files + compile step.

jimmy-zendar commented on raw_radar_reader/README.md:
```diff
@@ -0,0 +1,72 @@
+protoc radar.proto --python_out .
+protoc acquisition.proto --python_out .
+cd ../
+export PYTHONPATH=${PYTHONPATH}:<path-to-ZendarSDK>/raw_radar_reader/protocol
```
can this not be added into the environment.sh script such that we have all documentation at one place?

antoniopug replied:
Good idea, I added a environment.sh specific to this project (since it's standalone).

jimmy-zendar reviewed (approved):
I think this is fine if we want to just release the data. This code of course does not mingle with existing code and there are quite a bit of duplication in functionality. I think in the not distance future we will need to clean all this up so our python codes are coherent.

osamasal commented:
@antoniopug one issue that @plb9772 brought up is that we have moved all protos to ```ZenCode``` and we're in the process of publishing public protos via .deb packages in ```ZendarSDK```. This PR adds protos back to ```ZendarSDK```, but it would be best if we rely on the .deb packages to distribute the protos. If this PR can wait for the .deb package to be merged into ```ZendarSDK```, then that would be best, otherwise, we can always remove them later after the .deb packages are merged into ```ZendarSDK```

antoniopug commented:
@osamasal yeah would love to move over to the deb thing asap. When is that expected to land?

antoniopug committed:
Address review comments

antoniopug committed:
Address revie comments

plb9772 commented:
I know I was not asked to review this but I'd like to throw some thoughts here:
- I find the structure of the repo that way not super clear. There are  `python/viewer` and `raw_data_reader/python`. I understand that we may want to have two different requirements and setup for each, but so in that case, I think it would be clearer to have `python/viewer` and `python/raw_data_reader` with each of them having a README with the Setup instructions and explanations/quick example
- However, I am not totally sure we want to have the two separated in term of setup, since they will have in common:
   - long-term, we want them to both rely on the protobufs of the debian packages
   - there is the `python/third_party` folder that a user may want to use whatever which application is interesting to him
   - It seems that they are some similarities between `ProtoStreamReader` and what is in `acquisition.py`, both allowing to read protobuf. This is the kind of "duplication" I feel can become dangerous. Since `ProtoStreamReader` seems more "root", I would have it a parent of what is in `acquisition.py` (as much as possible). Furthermore, `ProtoStreamReader` could be used to describe an easy way for a user to build its own reader of protobuf (I think it would be of great value, considering the question asked in issue and some discussion I had with people at Joby). We could have also `python/protobuf_reader`that has a little documentation on how to do so (with the code of `ProtoStreamReader/Writer` that the two others applications would import).

This is just some quick thoughts after reading quickly this PR, I might miss some broader picture here.

antoniopug commented:
@plb9772 Thanks for your comments. I agree with what you're saying, and I'm not that happy at doing this tool this way.

The goal here was to quickly get this particular tool out the door to one collaborating group (U-Alabama). We need to have a discussion around what are the anticipated use cases for collaborating with external groups, and what is the right architecture for ZendarSDK to best support those use cases that we choose. In the absence of that clarity and infrastructure, I had to just build a first pass at this that is very specific to one particular use case.

osamasal commented:
> @osamasal yeah would love to move over to the deb thing asap. When is that expected to land?

Should be very soon. I have the [PR](https://github.com/ZendarInc/ZendarSDK/pull/27) out for review and I'm just waiting for sign-off before merging. In this PR, I updated the CPP examples, but I'll update and fix the existing Python tools under ```ZendarSDK/python/viewer``` shortly after. My ETA for merging this PR is early week of 03/07/2021.

osamasal commented on raw_radar_reader/README.md:
````diff
@@ -0,0 +1,60 @@
+
+To use these tools, perform the following steps:
+```
+source environment.sh
````
Once https://github.com/ZendarInc/ZendarSDK/pull/27 is merged, we will need to provide instructions on generating the Python ```_pb2``` files from the protobuf files. We need this for ```zendar-zenapi_amd64.deb``` since this is the package containing all ```.proto``` files.

osamasal replied:
> > @osamasal yeah would love to move over to the deb thing asap. When is that expected to land?
> 
> Should be very soon. I have the [PR](https://github.com/ZendarInc/ZendarSDK/pull/27) out for review and I'm just waiting for sign-off before merging. In this PR, I updated the CPP examples, but I'll update and fix the existing Python tools under `ZendarSDK/python/viewer` shortly after. My ETA for merging this PR is early week of 03/07/2021.

@antoniopug https://github.com/ZendarInc/ZendarSDK/pull/27 has been merged, so the protos are now part of the ```zendar-zenapi.deb``` package.

antoniopug commented:
Rebased on `main` to take in the deb packages. Will now work those in.

antoniopug committed:
Use the deb proto files

antoniopug commented:
@osamasal I removed the proto files, updated the README to use the build in `ZendarSDK/cpp/`, and updated the environment.sh script as well. 

I tested this by zero-ing out my PYTHONPATH and following the steps in the README, and I was able to run the example.py script without any issues. For the `--calibration` flag you can point it to `ZenCode/ZenModel/radar/antenna/intrinsic/ZRCA0003.pt`

Can you review the changes and re-try the build from scratch to check that it all works?

osamasal commented:
> @osamasal I removed the proto files, updated the README to use the build in `ZendarSDK/cpp/`, and updated the environment.sh script as well.
> 
> I tested this by zero-ing out my PYTHONPATH and following the steps in the README, and I was able to run the example.py script without any issues. For the `--calibration` flag you can point it to `ZenCode/ZenModel/radar/antenna/intrinsic/ZRCA0003.pt`
> 
> Can you review the changes and re-try the build from scratch to check that it all works?

Will do!

antoniopug committed:
Move raw-radar-reader under python

antoniopug commented:
I moved `raw_radar_reader` under `python` per PL's suggestion. Also adding him as a reviewer.

I have tested the new instructions and they seem to work.

osamasal commented:
> > @osamasal I removed the proto files, updated the README to use the build in `ZendarSDK/cpp/`, and updated the environment.sh script as well.
> > I tested this by zero-ing out my PYTHONPATH and following the steps in the README, and I was able to run the example.py script without any issues. For the `--calibration` flag you can point it to `ZenCode/ZenModel/radar/antenna/intrinsic/ZRCA0003.pt`
> > Can you review the changes and re-try the build from scratch to check that it all works?
> 
> Will do!

OK, I just tried your instructions and all looks good. I got a bunch of these warning messages, but I assume this is OK:

```
WARN: Frame out of sync:  22782
WARN: Frame out of sync:  22782
WARN: Frame out of sync:  22782
WARN: Frame out of sync:  22782
WARN: Frame out of sync:  22782
WARN: Frame out of sync:  22782
WARN: Frame out of sync:  22782
```
Also, I'm attaching the image that came up after running the command on the ```zen-ffa/nyquist/2020-12-04_0048/000``` data set. 
![Screenshot from 2021-03-10 11-43-37](https://user-images.githubusercontent.com/23441392/110688288-435a4900-8196-11eb-88af-f5b0f78933f8.png)

antoniopug commented:
@osamasal that all looks correct, thanks for your help reviewing! I think that with the repo refactoring I did this morning (at yours and PL's suggestion), it's looking like a lot cleaner integration of this tool.

osamasal commented:
> @osamasal that all looks correct, thanks for your help reviewing! I think that with the repo refactoring I did this morning (at yours and PL's suggestion), it's looking like a lot cleaner integration of this tool.

:+1:

plb9772 commented on python/README.md:
````diff
@@ -18,12 +22,23 @@ with the Zendar data
+
+When you open a new terminal, set your environment variables using this script:
+```
+source ../environment.sh
````
I am very picky here, but since the `environment.sh` is only needed for python, wouldn't it make sense to move it under `/python` and update the documentation here consequently ?

antoniopug replied:
Yes, done.

plb9772 commented on python/raw_radar_reader/README.md:
````diff
@@ -0,0 +1,48 @@
+## Example Script
+
+```
+python3 ./python/example.py --root <path-to-data> --calibration <path-to-cal-file>
````
With the moving around of the file, it should be `python3 example.py` only.

antoniopug replied:
Thanks for finding that

plb9772 reviewed (approved):
I also really like the refactoring under `/python` of the tools, I think the repo is in a pretty good shape now. I made a few minor comments.
Being picky until the end, I think it would be good to add something pointing to your tool in the REDAME at the root, as it is for `image_viewer`

I have already mentionned that earlier, but I think there is some refactoring to do in the way we read the protobuf between the two tools (between `raw_radar_reader/acquisition.py`, `raw_radar_reader/reader.py` and `viewer/radar_data_streamer.py`). As you said it should not block this PR but you may want to create a ticket to track it.

antoniopug committed:
Address review comments

antoniopug committed:
Also update main README
